src/thread/queue.cr

```diff
@@ -0,0 +1,53 @@
+# Based on rubysl-thread ruby gem implementation
+# Copyright (c) 2013, Brian Shirai
+# Licensed under the BSD 3-clause
+
+# :nodoc:
+class Thread(T, R)
+  # :nodoc:
+  class Queue(T)
+    class Error < Exception; end
+
+    def initialize
+      @que = Deque(T).new(16)
+      @mutex = Thread::Mutex.new
```

Comment:
For example try to lock with `cmpxchg(counter, 1, 0)` and yielding the fiber until it succeeds, then release the lock with `modifyrmw(counter, sub, counter, 1)`. We'd need to differentiate between pushing from the event-loop thread (there is a fiber) and calling from a thread (no fiber) and act accordingly by maybe using a semaphore from the thread instead of yielding.

This may be too complex, thought, and we should just not care until we tackle parallelism.